core/src/test/scala/unit/kafka/zookeeper/ZooKeeperClientTest.scala

```diff
@@ -385,6 +385,61 @@ class ZooKeeperClientTest extends ZooKeeperTestHarness {
     } finally zooKeeperClient.close()
   }
 
+  /**
+    * Tests that if session expiry notification is received while a thread is processing requests,
+    * session expiry is handled and the request thread completes with responses to all requests,
+    * even though some requests may fail due to session expiry or disconnection.
+    */
+  @Test
+  def testSessionExpiry(): Unit = {
+    val maxInflightRequests = 2
+    val requestThreadSemaphore = new Semaphore(0)
+    val sessionExpireSemaphore = new Semaphore(maxInflightRequests + 1) // expire after a few requests
+    val sendSize = maxInflightRequests * 50
+    @volatile var resultCodes: Seq[Code] = null
+    val zooKeeperClient = new ZooKeeperClient(zkConnect, zkSessionTimeout, zkConnectionTimeout, maxInflightRequests,
+      time, "testGroupType", "testGroupName") {
+      override def send[Req <: AsyncRequest](request: Req)(processResponse: Req#Response => Unit): Unit = {
+        requestThreadSemaphore.release()
+        super.send(request)(processResponse)
+        sessionExpireSemaphore.acquire()
+      }
+    }
+    try {
+      val requestThread = new Thread {
+        override def run(): Unit = {
+          val requests = (1 to sendSize).map(i => GetDataRequest(s"/$i"))
+          resultCodes = zooKeeperClient.handleRequests(requests).map(_.resultCode)
+        }
+      }
+      requestThread.start()
+      requestThreadSemaphore.acquire() // Wait for request thread to start processng requests
+      sessionExpireSemaphore.release(sendSize) // Resume request thread before expiring session
+
+      // Trigger session expiry by reusing the session id in another client
+      val dummyWatcher = new Watcher {
+        override def process(event: WatchedEvent): Unit = {}
+      }
+      val anotherZkClient = new ZooKeeper(zkConnect, 1000, dummyWatcher,
+        zooKeeperClient.currentZooKeeper.getSessionId,
+        zooKeeperClient.currentZooKeeper.getSessionPasswd)
+      assertNull(anotherZkClient.exists("/nonexistent", false)) // Make sure new client works
+      anotherZkClient.close()
```

Comment:
Hmm, since we released sessionExpireSemaphore, all requests could have been completed before the close() call. Not sure how we can guarantee that some requests will see Code.SESSIONEXPIRED.

Reply:
@junrao Thanks for the review. I was thinking that it would be very difficult to ensure that some requests are failed with `SESSIONEXPIRED`, so I went with a test with a large enough number of requests to make it work.  I have updated the test to make it more stable. We should see both `CONNECTIONLOSS` and `SESSIONEXPIRED` now, but I made the test more tolerant by asserting on just one of these combined with an assertion that session expiry event was handled.